### PR TITLE
Oracle producer add missing functions

### PR DESCRIPTION
**Pull request by hazardv**
Adds functionality to Producer/Oracle.pm to:
- Alter create constraints
- Alter drop constraints
- Drop fields
- Drop tables
- Alter create index
- Alter drop index

Related tests have been added to prove changes and all existing tests pass.
Automatic creation of trigger to insert sysdate to all timestamp fields has been commented out. That should not be done automatically. 

Changes were modeled after the structure of Producer/MySQL.pm.

This branch also includes changes from pull request #142

### CODE REVIEW

File: lib/SQL/Translator/Producer/Oracle.pm [hunk 1]
```diff
@@ -390,23 +391,9 @@ sub create_table {
...
-            elsif ( $index_type eq UNIQUE ) {
-                $index_name = $index_name ? mk_name( $index_name )
-                    : mk_name( $table_name, $index_name || 'i' );
-                $index_name = quote($index_name, $qf);
-                push @index_defs,
-                    "CREATE UNIQUE INDEX $index_name on $table_name_q (".
-                        join( ', ', @fields ).
-                    ")$index_options";
+            elsif ($index_type eq NORMAL or $index_type eq UNIQUE) {
+                warn "CAlling create index with options: " . Dumper($options) . "\n";
```
Comment (mohawk2) [inline lib/SQL/Translator/Producer/Oracle.pm]: "CAlling" typo

File: lib/SQL/Translator/Producer/Oracle.pm [hunk 2]
```diff
@@ -668,19 +655,20 @@ sub create_field {
...
-          "BEFORE INSERT OR UPDATE ON $table_name_q\n".
-          "FOR EACH ROW WHEN (new.$field_name_q IS NULL)\n".
-          "BEGIN\n".
-          " SELECT sysdate INTO :new.$field_name_q FROM dual;\n".
-          "END;\n";
-
-          push @trigger_defs, $trigger;
-    }
+    # Do not create a trigger to insert sysdate to all timestamp fields
+    # if ( lc $field->data_type eq 'timestamp' ) {
```
Comment (mohawk2) [inline lib/SQL/Translator/Producer/Oracle.pm:622]: I don't think it's good practice to leave slabs of commented code. If you want to delete it, delete it.

File: t/54-oracle-sql-diff.t [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    target_schema => $t->schema,
+    sqlt_args => {quote_identifiers => 1}
+   });
+
+
+my $diff = $d->compute_differences->produce_diff_sql || die $d->error;
+
+ok($diff, 'Diff generated.');
+
+warn "The diff is: " . Dumper($diff);
```
Comment (mohawk2) [inline t/54-oracle-sql-diff.t]: I don't think `warn` in tests is good practice?

**Review by mohawk2 — CHANGES_REQUESTED** [review]

File: lib/SQL/Translator/Producer/Oracle.pm (hunk 1 above)
Comment (hazardv) [reply lib/SQL/Translator/Producer/Oracle.pm]: My bad, that warn was there for debugging. I missed it when I was cleaning up. I'll take that out.

File: lib/SQL/Translator/Producer/Oracle.pm (hunk 2 above)
Comment (hazardv) [reply lib/SQL/Translator/Producer/Oracle.pm:622]: I agree, I just wasn't sure how things were being handled on this project. I wanted that change and its reason to be obvious. I will get rid of that.

File: t/54-oracle-sql-diff.t (hunk 3 above)
Comment (hazardv) [reply t/54-oracle-sql-diff.t]: Again, My bad, that was supposed to get removed. I will take care of it.

**Comment by hazardv** [comment]
@mohawk2 The code you reviewed was from a few commits back. This is the first time I am attempting to contribute to a project like this so I am betting that I messed a few things up. 

I have been using DBIx::Class and DBIx::Class::Schema::Versioned for another project and while working on that I discovered functionality that was missing from the Oracle producer. I added in the missing functionality and submitted the pull request. I continued with my project, made changes to the data model, attempted to generate the DDL, and discovered a few more things missing from the Oracle producer. So I added those things into the same branch because it seemed reasonable and pushed the changes. I did that whole process one more time and now I am like 97% sure that all of the missing functionality has been implemented and I won't make any more changes to the branch. 

Please check out the latest commit [b14da74](https://github.com/dbsrgits/sql-translator/pull/143/commits/b14da74a5bb7e2e342e2651c6abfdf656e153128) and let me know if you see any issues.

File: lib/SQL/Translator/Producer/Oracle.pm (hunk 2 above)
Comment (hazardv) [reply lib/SQL/Translator/Producer/Oracle.pm]: Fixed in commit [b14da74](https://github.com/dbsrgits/sql-translator/pull/143/commits/b14da74a5bb7e2e342e2651c6abfdf656e153128)

File: lib/SQL/Translator/Producer/Oracle.pm (hunk 1 above)
Comment (hazardv) [reply lib/SQL/Translator/Producer/Oracle.pm]: That is from an old commit.

**Comment by mohawk2** [comment]
Thanks for the updates.

The current list of commits sort of meanders, which isn't a criticism, but does highlight a possible process-improvement, especially to help reviewers. Since this is on a branch belonging to you, it's safe to force-push it. The way I work is to end up with the branch I'm making having a series of small commits, each of which meaningfully transforms the starting code towards the desired end-state. This often involves squashing together (using `rebase` "fixup" command) previous versions of a change, with a correction. This means the final list of commits doesn't have things like commented code, then removal of that (as here). Having adjusted the list of commits to a new one, you'd then need to force-push it.

The alternative (which isn't fatal) is that this set of commits will either (if merged in as-is) be slightly untidy for future people looking to see why a piece of code is as it is, or (if squash-merged) a single quite large commit which can take longer to comprehend.

Tagging in @rabbiveesh.

**Comment by hazardv** [comment]
@mohawk2 Thank you for the pointers. I will try to clean up my commits to be more concise in the future. I found one last issue with altering constraints for Oracle while working on my personal project (dropping an unnamed primary key constraint) and have pushed the change to this branch. At this point, I have personally tested all of the constraint alteration functionality in my other project (as well as adding tests for this project) and am confident in them.

File: lib/SQL/Translator/Diff.pm [hunk 4]
```diff
@@ -92,7 +92,7 @@ sub schema_diff {
     $options ||= {};
 
     my $obj = SQL::Translator::Diff->new( {
-      %$options,
+      sqlt_args => $options,
```
Comment (rabbiveesh) [inline lib/SQL/Translator/Diff.pm:95]: hey, why did you change this line?
It looks to me like the way it's meant to be used is to passthrough args to SQLT::Diff, one arg of which is the args which are passed down to the inner SQLT instance

File: lib/SQL/Translator/Parser/Oracle.pm [hunk 5]
```diff
@@ -489,7 +489,17 @@ parens_name_list : '(' NAME(s /,/) ')'
 field_meta : default_val
     | column_constraint
 
-default_val  : /default/i VALUE
+default_val  : 
+    /default/i CURRENT_TIMESTAMP
```
Comment (rabbiveesh) [inline lib/SQL/Translator/Parser/Oracle.pm:493]: just to get this straight: this is special casing CURRENT_TIMESTAMP b/c it's a magic constant in the Oracle grammar, and you don't want SQLT to quote it down the line?

If that's the case, I don't know if this is the right place to take care of special casing it here. If you look at the SQLite parser, they special case it in the rule for `VALUE`, which makes more sense to me.

Comment (rabbiveesh) [reply lib/SQL/Translator/Parser/Oracle.pm:493]: ah, looking at it again, I see that's exactly what the MySQL parser does. So then I'm okay w/ this, though I think the SQLite way makes more sense

**Review by rabbiveesh — CHANGES_REQUESTED** [review]
Hey, I only ran through a little bit, but then I saw that tests are failing on my box.
I suggest you make sure that all tests pass.

Here's the list of failing files:
t/54-oracle-sql-diff.t      
t/30sqlt-new-diff-mysql.t   
t/30sqlt-new-diff-pgsql.t   
t/30sqlt-new-diff-sqlite.t  
t/51-xml-to-oracle.t        
t/51-xml-to-oracle_quoted.t

File: lib/SQL/Translator/Parser/Oracle.pm (hunk 5 above)
Comment (hazardv) [reply lib/SQL/Translator/Parser/Oracle.pm:493]: I handled it that way because of how it was handled for MySQL and CURRENT_TIMESTAMP is handled the same way by Oracle and MySQL databases. This is my first time looking at the parser syntax and its quite confusing so I tried to keep the changes as simple and as close to what I could understand as possible.

File: lib/SQL/Translator/Diff.pm (hunk 4 above)
Comment (hazardv) [reply lib/SQL/Translator/Diff.pm]: At this point, I am not quite sure why I did that. I have been making all of these changes as a result of trying to get a project off the ground using DBIx::Class with an Oracle DB and either DBIx::Class::Schema::Versioned or DBIx::Class::DeploymentHandler for deployment. I know at one point I was running into an issue where schema_diff was getting called directly by a module and when that was happening the sqlt_args that I had passed to the module were getting lost. Changing that line fixed the issue but I hadn't realized how many other things it actually broke. I reverted that and everything in my code seems to be working properly so I am guessing that was related to an issue I was having with DBIx::Class::DeploymentHandler. If I run into the issue again I will do more thorough testing and open a new issue.

**Comment by hazardv** [comment]
@rabbiveesh after reverting the change in Diff.pm (and fixing one of my test plans) all of the tests are passing on my box except for t/60roundtrip.t. It is looking for t/data/roundtrip_autogen.yaml which I am unable to locate in my branch or in the current project master.

File: lib/SQL/Translator/Diff.pm (hunk 4 above)
Comment (rabbiveesh) [reply lib/SQL/Translator/Diff.pm]: k, great

**Comment by rabbiveesh** [comment]
Yeah, to get the roundtrip test working you need some script. You can run Makefile.PL, and then run `make test`, it should initialize whatever you need.

**Review by rabbiveesh — CHANGES_REQUESTED** [review]
could you just remove the debug statements from the Oracle producer?

I mean, unless you think we should keep them?

**Comment by hazardv** [comment]
@rabbiveesh I added them following how they were used in the MySQL producer, tagged them with "ORA" at the start, and included them sparingly figuring they could be useful for someone debugging their code in the future. Since they only generate output if the user specifies `$DEBUG = 1` I didn't think they hurt anything to be there but if you want them removed I will remove them.

**Review by rabbiveesh — APPROVED** [review]
K, wasnt sure if you had missed some in the last commit.
@mohawk2 any thoughts?

**Comment by hazardv** [comment]
Hey @mohawk2, can you please take a look at this when you get a minute?

**Comment by hazardv** [comment]
@mohawk2 Happy New Year!!!

**Review by mohawk2 — COMMENTED** [review]
I'm a bit surprised by the tests being removed from the XML-related tests. If @rabbiveesh is content it's for a good reason then so am I. Otherwise, all looks fine.

**Comment by hazardv** [comment]
> I'm a bit surprised by the tests being removed from the XML-related tests. If @rabbiveesh is content it's for a good reason then so am I. Otherwise, all looks fine.

@mohawk2  The only things I removed from the XML tests were the checks looking for the triggers that previously got created for all timestamp fields. Since there is no reason to create that trigger for all timestamp fields I commented out the code in commit https://github.com/dbsrgits/sql-translator/pull/143/commits/ab42ef849111e4091ccae0981bb4241d60c15146. I then removed the commented-out code in commit https://github.com/dbsrgits/sql-translator/pull/143/commits/b14da74a5bb7e2e342e2651c6abfdf656e153128.

**Comment by rabbiveesh** [comment]
sorry for the delay on this; we're good to go
